server/src/internalClusterTest/java/org/elasticsearch/search/SearchCancellationIT.java

```diff
@@ -270,29 +242,60 @@ public void testCancelFailedSearchWhenPartialResultDisallowed() throws Exception
             );
             assertThat(e.getMessage(), containsString("Partial shards failure"));
         });
+
+        // When the search request executes, block all shards except 1.
+        final List<SearchShardBlockingPlugin> searchShardBlockingPlugins = initSearchShardBlockingPlugin();
+        AtomicBoolean letOneShardProceed = new AtomicBoolean();
+        CountDownLatch shardTaskLatch = new CountDownLatch(1);
+        for (SearchShardBlockingPlugin plugin : searchShardBlockingPlugins) {
+            plugin.setRunOnNewReaderContext((ReaderContext c) -> {
+                System.out.println("blah");
+                if (letOneShardProceed.compareAndSet(false, true)) {
+                    // Let one shard continue.
+                } else {
+                    try {
+                        shardTaskLatch.await(); // Bock the other shards.
+                    } catch (InterruptedException e) {
+                        throw new AssertionError(e);
+                    }
+                }
+            });
+        }
+
+        // For the shard that was allowed to proceed, have a single query-execution thread throw an exception.
+        final List<ScriptedBlockPlugin> plugins = initBlockFactory();
+        AtomicBoolean oneThreadWillError = new AtomicBoolean();
+        for (ScriptedBlockPlugin plugin : plugins) {
+            plugin.disableBlock();
+            plugin.setBeforeExecution(() -> {
+                if (oneThreadWillError.compareAndSet(false, true)) {
+                    throw new IllegalStateException("This will cancel the ContextIndexSearcher.search task");
+                }
+            });
+        }
+
+        // Now run the search request.
         searchThread.start();
+
         try {
-            assertBusy(() -> assertThat(getSearchTasks(), hasSize(1)));
-            queryLatch.countDown();
             assertBusy(() -> {
-                final List<SearchTask> searchTasks = getSearchTasks();
-                // The search request can complete before the "cancelledLatch" is latched if the second shard request is sent
-                // after the request was cancelled (i.e., the child task is not allowed to start after the parent was cancelled).
-                if (searchTasks.isEmpty() == false) {
-                    assertThat(searchTasks, hasSize(1));
-                    assertTrue(searchTasks.get(0).isCancelled());
+                final List<SearchTask> coordinatorSearchTask = getCoordinatorSearchTasks();
+                assertThat("The Coordinator should have one SearchTask.", coordinatorSearchTask, hasSize(1));
+                assertTrue("The SearchTask should be cancelled.", coordinatorSearchTask.get(0).isCancelled());
+                for (var shardQueryTask : getShardQueryTasks()) {
+                    assertTrue("All SearchShardTasks should then be cancelled", shardQueryTask.isCancelled());
```

Comment:
This ensures that the cancelation propagated up from the thread failure and was fanned-out to the other shards.